### PR TITLE
[ci] #2461: Improve iroha2 CI

### PR DESCRIPTION
**Pull request by BAStos525**
Signed-off-by: BAStos525 <jungle.vas@yandex.ru>

### Description of the Change

1. Move `iroha2:base` image build to the separate workflow and publish it on Dockerhub in order to make it reusable.
2. Distribute `iroha2:base` and `iroha2:build` images usage among branches.
3. Update references of `building load-rs script` jobs to the [iroha2-longevity-load-rs repository](https://github.com/soramitsu/iroha2-longevity-load-rs.git).
4. Switch to `actions/checkout@v3` where it hasn't been done so far.
5. Add step to trigger workflow of `Cargo.toml` dependencies updating.

### Issue
1. This PR resolves the major part of #2461 issue.
2. It also resolves [this issue](https://github.com/soramitsu/iroha2-longevity-load-rs/issues/18) partially.

### Benefits
1. Make iroha2 CI faster and more effective.
2. Make `iroha2-ci:<tag>` and `iroha2:build-<tag>` images be available for calling from other branches.
3. Follow to the last versions of third-party actions.
6. Update workflows calling for `longevity-load-rs` repo.

### Possible Drawbacks
1. A fresh feature: CI might be failed on any of `release` workflows or deploying `iroha2-ci` image.
2. It will not work until the corresponding `Dockerfile.base` and `Dockerfile.base` will be added to `iroha2` and `iroha2-lts` branches.
3. This [note](https://github.com/hyperledger/iroha/issues/2461#issuecomment-1184150212) hasn't still resolved (@s8sato @appetrosyan). [Original link to issue](https://github.com/hyperledger/iroha/pull/2433#discussion_r919960840).
4. `hyperledger/iroha2-ci` image doesn't compatible to be used for any of iroha2 tests workflows and jobs instead of `7272721/i2-ci:latest`. I spent some time to adopt it but it was unsuccessful due to a bunch of possible reasons. So, It is not the time for that yet.
5. Firstly we should merge [this PR](https://github.com/soramitsu/iroha2-longevity-load-rs/pull/23).

### CODE REVIEW

**Comment by s8sato** [comment]
> 3. This [note](https://github.com/hyperledger/iroha/issues/2461#issuecomment-1184150212) hasn't still resolved (@s8sato @appetrosyan). [Original link to issue](https://github.com/hyperledger/iroha/pull/2433#discussion_r919960840).

I'll address this in https://github.com/hyperledger/iroha/issues/2309#issuecomment-1192732181

**Comment by BAStos525** [comment]
One thing: since you are going to use a special `iroha2-ci:stable` and `iroha2-ci:lts` images sometimes, it means that `Dockerfile.base` and `Dockerfile.build` will be different for iroha2-dev, iroha2 and iroha-lts branches. I guess it's not convenient and it will cause a conflicts while merging branches.

File: .github/workflows/iroha2-deploy-base-image.yml
Comment (appetrosyan) [inline .github/workflows/iroha2-deploy-base-image.yml:2]: We should have a way of pushing the iroha2-ci:dev image before merging into iroha2-dev.

Comment (appetrosyan) [reply .github/workflows/iroha2-deploy-base-image.yml:2]: created issue

**Review by appetrosyan — APPROVED** [review]
LGTM.

**Comment by appetrosyan** [comment]
> One thing: since you are going to use a special iroha2-ci:stable and iroha2-ci:lts images sometimes, it means that Dockerfile.base and Dockerfile.build will be different for iroha2-dev, iroha2 and iroha-lts branches. I guess it's not convenient and it will cause a conflicts while merging branches.

We can work around the inconvenience if we query the context for the branch name. 

So on iroha2 it resolves to iroha2:stable and so on.

File: .github/workflows/iroha2-deploy-base-image.yml
Comment (BAStos525) [reply .github/workflows/iroha2-deploy-base-image.yml:2]: DevOps team got creds for the user that has write access to Hyperledger DockerHub account.
[iroha2-ci:dev image](https://hub.docker.com/layers/iroha2-ci/hyperledger/iroha2-ci/dev/images/sha256-776f52d9c1247250c84c3c00b8f2465146e65571fe6fc76177db572f6d1bde30?context=explore)

File: .github/workflows/iroha2-dev-pr.yml
Comment (s8sato) [inline .github/workflows/iroha2-dev-pr.yml:28]: The same for other 16 lines in 7 files

File: .github/workflows/iroha2-deploy-base-image.yml
Comment (s8sato) [inline .github/workflows/iroha2-deploy-base-image.yml:4]:
```suggestion
on:
  workflow_dispatch:
  push:
```
Just in case `archlinux:base-devel` is silently outdated

File: .github/workflows/iroha2-deploy-base-image.yml
Comment (s8sato) [inline .github/workflows/iroha2-deploy-base-image.yml:81]:
```suggestion
jobs:
  base_image_deploy:
    strategy:
      matrix:
        tags: [dev, stable, lts]
    if: github.ref == 'refs/heads/iroha2-${{ matrix.tags }}'
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - name: Set up buildx
        uses: docker/setup-buildx-action@v2
        id: buildx
        with:
          install: true
      - name: Login to DockerHub
        uses: docker/login-action@v2
        with:
          username: ${{ secrets.DOCKERHUB_USERNAME }}
          password: ${{ secrets.DOCKERHUB_TOKEN }}
      - name: Build and push iroha2-ci:${{ matrix.tags }} image
        uses: docker/build-push-action@v3
        with:
          push: true
          tags: hyperledger/iroha2-ci:${{ matrix.tags }}
          context: ${{ env.REPOSITORY_URL }}#iroha2-${{ matrix.tags }}
          file: Dockerfile.base
          build-args: TAG=${{ matrix.tags }}
```

To apply this, we need to unify `iroha2` and `iroha2-stable` with `iroha2-stable`. What do you think @appetrosyan

File: Dockerfile.build
Comment (s8sato) [inline Dockerfile.build:1]: Since this docker file is also used for other tags than `dev`, it would better to remove the `dev` designation:

```suggestion
ARG TAG
FROM    hyperledger/iroha2-ci:$TAG
```

File: .github/workflows/iroha2-deploy-base-image.yml
Comment (appetrosyan) [reply .github/workflows/iroha2-deploy-base-image.yml:81]: Technically the `stable` `lts` and `dev` are all tags, not branches. Unification might be a good idea.

File: Dockerfile.build
Comment (appetrosyan) [reply Dockerfile.build:1]: Excellent!

File: .github/workflows/iroha2-lts.yml
Comment (appetrosyan) [inline .github/workflows/iroha2-lts.yml:18]: Good riddance

File: .github/workflows/iroha2-lts.yml
Comment (appetrosyan) [inline .github/workflows/iroha2-lts.yml:24]: Does this result in an extra `git clone`?

File: .github/workflows/iroha2.yml
Comment (s8sato) [inline .github/workflows/iroha2.yml:6]: If we synchronize tags and branches and unify the `iroha2` and `iroha2-stable` __branches__ as I mentioned in https://github.com/hyperledger/iroha/pull/2576#discussion_r937453231,

```suggestion
on:
  push:
    branches: [iroha2-stable]
```

Also, the workflow filenames will be renamed to `iroha2-stable.yml`, etc.

File: .github/workflows/iroha2-deploy-base-image.yml
Comment (s8sato) [reply .github/workflows/iroha2-deploy-base-image.yml:81]: Let me make a clarification. I'm suggesting to synchronize tags and branches.

image
image

So these branches will be `iroha2-stable`

Comment (s8sato) [reply .github/workflows/iroha2-deploy-base-image.yml:81]: I think this is a question of which tag we prefer, implicit `latest` or explicit `stable`

Comment (BAStos525) [reply .github/workflows/iroha2-deploy-base-image.yml]: I'll rename `iroha2` to `iroha2-stable` branch if so

File: .github/workflows/iroha2-lts.yml
Comment (BAStos525) [reply .github/workflows/iroha2-lts.yml:24]: It seems [like that](https://docs.docker.com/engine/reference/commandline/build/)

Comment (s8sato) [reply .github/workflows/iroha2-lts.yml:24]: I think `.` implies `${{ env.REPOSITORY_URL }}#branch`, thanks to `actions/checkout`

File: docker-compose.yml
Comment (s8sato) [inline docker-compose.yml:19]: What is the reason to comment out? This reverts the result of #2311

Comment (BAStos525) [reply docker-compose.yml:19]: I apologize, it's was caused by playing #2594 (I forgot to change the branch).

**Review by s8sato — APPROVED** [review]

**Review by appetrosyan — APPROVED** [review]

File: .github/workflows/iroha2-lts.yml
Comment (s8sato) [reply .github/workflows/iroha2-lts.yml:24]: Taken over by https://github.com/hyperledger/iroha/commit/3796924f02191981ff7cef3ceec4a6343874a95f in #2427

**Pull request merged**
